File: create_prompt.R
#!/usr/bin/env Rscript
# This R script scans your project directory and creates a file (output.txt) with your codebase
# that contains:
#   1. A tree-like ASCII representation of your project's directory structure.
#   2. The content of each code file (filtered by file type).
#
# This is particularly useful for providing context about your codebase to an AI or a human collaborator.
#
# Usage:
#   - Place this script in your project's root directory.
#   - Run it in the terminal with: ./create_prompt.R
#
# Configuration:
#   - EXCLUDED_DIRS: Directories to ignore (e.g., .git, node_modules).
#   - EXCLUDED_FILES: Specific files to ignore.
#   - EXCLUDED_EXTENSIONS: File extensions to ignore (e.g., images, JSON, etc.).
#   - BASE_DIRECTORY: The directory to scan (default is the current directory).



#  Do you like it? Add me on LinkedIn (Piet Jonker) 
#  Do you know about opportunities for AI/LLM development? --> email to piet@backlight.ai



# --- Configuration Section ---
EXCLUDED_DIRS <- c(".git", ".next", "node_modules")
EXCLUDED_FILES <- c("package.json", "package-lock.json", "create_prompt.R", ".DS_Store", "README.md")
EXCLUDED_EXTENSIONS <- tolower(c(".png", ".jpg", ".jpeg", ".gif", ".ico",
                                 ".pdf", ".zip", ".tar", ".gz", ".mp3", ".mp4", ".csv",
                                 ".xlsx", ".json", ".enc"))
BASE_DIRECTORY <- "."

# --- Helper Function: generate_tree ---
generate_tree <- function(directory, prefix = "") {
  # Recursively builds an ASCII tree of the directory structure.
  lines <- character(0)
  # Get a sorted list of entries in the directory (excluding hidden files).
  entries <- sort(list.files(directory, all.files = FALSE, no.. = TRUE))
  filtered_entries <- character(0)
  
  # Filter out directories and files that are excluded.
  for (entry in entries) {
    full_path <- file.path(directory, entry)
    if (dir.exists(full_path)) {
      if (!(entry %in% EXCLUDED_DIRS)) {
        filtered_entries <- c(filtered_entries, entry)
      }
    } else {
      if (!(entry %in% EXCLUDED_FILES)) {
        ext <- paste0(".", tolower(tools::file_ext(entry)))
        if (!(ext %in% EXCLUDED_EXTENSIONS)) {
          filtered_entries <- c(filtered_entries, entry)
        }
      }
    }
  }
  
  n <- length(filtered_entries)
  # Build tree lines with connectors.
  for (i in seq_along(filtered_entries)) {
    entry <- filtered_entries[i]
    connector <- if (i == n) "└── " else "├── "
    line <- paste0(prefix, connector, entry)
    lines <- c(lines, line)
    full_path <- file.path(directory, entry)
    # If the entry is a directory, recursively generate its subtree.
    if (dir.exists(full_path)) {
      extension <- if (i == n) "    " else "│   "
      subtree <- generate_tree(full_path, paste0(prefix, extension))
      lines <- c(lines, subtree)
    }
  }
  return(lines)
}

# --- Build the Directory Tree ---
# Inform the user about the script's purpose.
cat("Welcome! This script scans your project and generates a Markdown overview.\n")
base_name <- basename(normalizePath(BASE_DIRECTORY))
# Start the Markdown content with a header and the base directory.
markdown_lines <- c("# Project file structure", paste("📁", base_name))
tree_lines <- generate_tree(BASE_DIRECTORY)
markdown_lines <- c(markdown_lines, tree_lines)

# --- Collect Code File Contents ---
code_content <- list()

collect_code_files <- function(directory) {
  # Recursively collects the content of code files that pass the filtering criteria.
  files <- list.files(directory, all.files = FALSE, no.. = TRUE)
  for (file in files) {
    full_path <- file.path(directory, file)
    if (dir.exists(full_path)) {
      if (!(file %in% EXCLUDED_DIRS)) {
        collect_code_files(full_path)
      }
    } else {
      if (!(file %in% EXCLUDED_FILES)) {
        ext <- paste0(".", tolower(tools::file_ext(file)))
        if (!(ext %in% EXCLUDED_EXTENSIONS)) {
          content <- tryCatch(
            readLines(full_path, warn = FALSE),
            error = function(e) NA
          )
          if (!is.na(content[1])) {
            relative_path <- substring(normalizePath(full_path),
                                       nchar(normalizePath(BASE_DIRECTORY)) + 2)
            code_content[[relative_path]] <<- paste(content, collapse = "\n")
          }
        }
      }
    }
  }
}

# Start collecting code files from the base directory.
collect_code_files(BASE_DIRECTORY)

# --- Write the Output to a Markdown File ---
output <- paste(markdown_lines, collapse = "\n")
output <- paste(output, "\n\n# Code files\n", sep = "")

# Append each file's content to the output.
for (relative_path in names(code_content)) {
  output <- paste(output, sprintf("\n## %s\n", relative_path), sep = "")
  output <- paste(output, "```\n", sep = "")
  output <- paste(output, code_content[[relative_path]], "\n", sep = "")
  output <- paste(output, "```\n", sep = "")
}

# Write the final output to "output.txt".
writeLines(output, "output.txt")
cat("Output successfully written to 'output.txt'.\n")

File: create_prompt.js
#!/usr/bin/env node
/**
 * This script scans your project directory and generates a Markdown file (output.txt)
 * that includes:
 *   1. An ASCII tree representation of your project's structure.
 *   2. The contents of code files (after filtering out unwanted file types).
 *
 * This is useful for providing an overview of your project to an AI or a human reviewer.
 *
 * Usage:
 *   - Ensure Node.js is installed.
 *   - Run this script with: node create_prompt.js
 *   - The output file "output.txt" will be created in the current directory.
 *
 * Configuration:
 *   - EXCLUDED_DIRS: Directories to ignore (e.g., .git, node_modules).
 *   - EXCLUDED_FILES: Specific files to ignore.
 *   - EXCLUDED_EXTENSIONS: File extensions to ignore (e.g., images, JSON, etc.).
 */


/**
 * Do you like it? Add me on LinkedIn (Piet Jonker) 
 * Do you know about opportunities for AI/LLM development? --> email to piet@backlight.ai
 */

const fs = require('fs');
const path = require('path');

// --- Configuration Section ---
// Define directories that should be skipped.
const EXCLUDED_DIRS = new Set(['.git', '.next', 'node_modules']);
// Define files that should be skipped.
const EXCLUDED_FILES = new Set(['package.json', 'package-lock.json', 'create_prompt.js', '.DS_Store', 'README.md']);
// Define file extensions to ignore.
const EXCLUDED_EXTENSIONS = new Set([
  '.png', '.jpg', '.jpeg', '.gif', '.ico',
  '.pdf', '.zip', '.tar', '.gz', '.mp3', '.mp4', '.csv',
  '.xlsx', '.json', '.enc'
]);
// Set the base directory for scanning.
const BASE_DIRECTORY = '.';

/**
 * Recursively generate a tree representation of the directory structure.
 * Uses ASCII connectors ("├── " and "└── ") to display hierarchy.
 *
 * @param {string} directory - The directory to scan.
 * @param {string} prefix - Indentation prefix for the current level.
 * @returns {string[]} Array of strings representing each line of the tree.
 */
function generateTree(directory, prefix = "") {
  let lines = [];
  let entries;
  try {
    // Read entries from the directory.
    entries = fs.readdirSync(directory);
  } catch (err) {
    // If the directory can't be read, return an empty list.
    return lines;
  }
  // Sort the entries alphabetically.
  entries.sort();

  // Filter out unwanted directories and files.
  let filteredEntries = entries.filter(entry => {
    const fullPath = path.join(directory, entry);
    let stat;
    try {
      stat = fs.lstatSync(fullPath);
    } catch (err) {
      return false;
    }
    if (stat.isDirectory()) {
      return !EXCLUDED_DIRS.has(entry);
    } else {
      if (EXCLUDED_FILES.has(entry)) return false;
      const ext = path.extname(entry).toLowerCase();
      return !EXCLUDED_EXTENSIONS.has(ext);
    }
  });

  // Build the tree lines using appropriate connectors.
  filteredEntries.forEach((entry, index) => {
    const isLast = index === filteredEntries.length - 1;
    const connector = isLast ? "└── " : "├── ";
    lines.push(prefix + connector + entry);
    const fullPath = path.join(directory, entry);
    // If the entry is a directory, recursively generate its subtree.
    if (fs.existsSync(fullPath) && fs.lstatSync(fullPath).isDirectory()) {
      const extension = isLast ? "    " : "│   ";
      lines = lines.concat(generateTree(fullPath, prefix + extension));
    }
  });
  return lines;
}

// --- Build the Directory Tree ---
// Inform the user about the script's purpose.
console.log("Welcome! This script scans your project and generates a Markdown overview.");
const baseName = path.basename(path.resolve(BASE_DIRECTORY));
// Initialize the Markdown content with a header and the base directory name.
let markdownLines = [`# Project file structure`, `📁 ${baseName}`];
// Generate the tree structure and append it.
markdownLines = markdownLines.concat(generateTree(BASE_DIRECTORY));

// --- Collect Code File Contents ---
// This object will map relative file paths to their content.
let codeContent = {};

/**
 * Recursively collects code files that pass filtering from the given directory.
 *
 * @param {string} directory - The directory to scan.
 */
function collectCodeFiles(directory) {
  let entries;
  try {
    entries = fs.readdirSync(directory);
  } catch (err) {
    return;
  }
  entries.forEach(entry => {
    const fullPath = path.join(directory, entry);
    let stat;
    try {
      stat = fs.lstatSync(fullPath);
    } catch (err) {
      return;
    }
    if (stat.isDirectory()) {
      // Recursively process subdirectories if not excluded.
      if (!EXCLUDED_DIRS.has(entry)) {
        collectCodeFiles(fullPath);
      }
    } else {
      // Process the file if it's not excluded.
      if (!EXCLUDED_FILES.has(entry)) {
        const ext = path.extname(entry).toLowerCase();
        if (!EXCLUDED_EXTENSIONS.has(ext)) {
          try {
            // Read the file content.
            const content = fs.readFileSync(fullPath, 'utf-8');
            const relativePath = path.relative(BASE_DIRECTORY, fullPath);
            codeContent[relativePath] = content;
          } catch (err) {
            console.error(`Warning: Could not read file ${fullPath}. Error: ${err}`);
          }
        }
      }
    }
  });
}
collectCodeFiles(BASE_DIRECTORY);

// --- Write the Output to a Markdown File ---
console.log("Writing output to 'output.txt'...");
let output = markdownLines.join('\n');
output += '\n\n# Code files\n';
// Append each file's content in a separate section.
for (const [relativePath, content] of Object.entries(codeContent)) {
  output += `\n## ${relativePath}\n\`\`\`\n${content}\n\`\`\`\n`;
}
try {
  fs.writeFileSync('output.txt', output, 'utf-8');
  console.log("Output successfully written to 'output.txt'.");
} catch (err) {
  console.error(`Error: Failed to write output file. ${err}`);
}
